core/arch/arm/mm/core_mmu.c

```diff
@@ -118,91 +254,153 @@ static struct map_area *find_map_by_pa(unsigned long pa)
 	return NULL;
 }
 
-static void insert_mmap(struct tee_mmap_region *mm, size_t max_elem,
-		struct tee_mmap_region *mme)
+extern const struct core_mmu_phys_mem __start_phys_mem_map_section;
+extern const struct core_mmu_phys_mem __stop_phys_mem_map_section;
+
+static void add_phys_mem(struct tee_mmap_region *memory_map, size_t num_elems,
+			 const struct core_mmu_phys_mem *mem, size_t *last)
 {
-	size_t n;
+	size_t n = 0;
+	paddr_t pa;
+	size_t size;
 
-	for (n = 0; n < (max_elem - 1); n++) {
-		if (!mm[n].size) {
-			mm[n] = *mme;
+	DMSG("%s %d 0x%08" PRIxPA " size 0x%08zx",
+	     mem->name, mem->type, mem->addr, mem->size);
+	while (true) {
+		if (n >= (num_elems - 1)) {
+			EMSG("Out of entries (%zu) in memory_map", num_elems);
+			panic();
+		}
+		if (n == *last)
+			break;
+		pa = memory_map[n].pa;
+		size = memory_map[n].size;
+		if (mem->addr >= pa && mem->addr <= (pa + (size - 1)) &&
+		    mem->type == memory_map[n].type) {
+			DMSG("Physical mem map overlaps 0x%" PRIxPA, mem->addr);
+			memory_map[n].pa = MIN(pa, mem->addr);
```

Comment:
What exactly is happening here? If we try to call `add_phys_mem` twice with a physical address in the same region and **but** with same type, then we just re-assign/update the region? I.e., as base pointer for that region we use the lowest address, but we extend the size to cover an eventual extended region? Did I get this or is my understanding completely wrong?